blog/_posts/2019-01-18-fluxdiffeq.md

```diff
@@ -0,0 +1,650 @@
+
+$$
+Put Lotka-Volterra equations
+$$
+
+and encoded in Julia like:
+
+```julia
+using DifferentialEquations
+function lotka_volterra(du,u,p,t)
+  x, y = u
+  α, β, δ, γ = p
+  du[1] = dx = α*x - β*x*y
+  du[2] = dy = -δ*y + γ*x*y
+end
+u0 = [1.0,1.0]
+tspan = (0.0,10.0)
+p = [1.5,1.0,3.0,1.0]
+prob = ODEProblem(lotka_volterra,u0,tspan,p)
+```
+
+Then to solve the differential equations, you can simply call `solve` on the
+`prob`:
+
+```julia
+sol = solve(prob)
+using Plots
+plot(sol)
+```
+
+![LV Solution Plot](https://user-images.githubusercontent.com/1814174/51388169-9a07f300-1af6-11e9-8c6c-83c41e81d11c.png)
+
+Optional arguments can be passed to control the ODE solver to customise most aspects of the numerical integration.
+For example, if you've some familiarity with the area, a good choice for
+this equation is to use the 5th order Runge-Kutta method `Tsit5()` (related to the `dopri` method,
+or `ode45` in MATLAB). We can also say that we want this solved with a relative error
+tolerance of `1e-8`, and that we want the solution saved at every `0.1` time points. This looks like:
+
+```julia
+sol = solve(prob,Tsit5(),saveat=0.1,reltol=1e-8)
+```
+
+There are a lot more arguments you can pass in to customise most aspects of the ODE solver,
+as
+[specified in the documentation](http://docs.juliadiffeq.org/latest/basics/common_solver_opts.html),
+but these will suffice for our demonstration.
+
+Note that the solution of a differential equation is a continuous object, so
+for example
+
+```julia
+sol(5.24)
+```
+
+gives the solution at time `5.24`. It also acts as an array, where
+
+```julia
+sol[2,5]
+sol.t[5]
+```
+
+is the solution of the 2nd variable (`y`) at the the 5th time point. Notice that
+the 5th time point is `0.4` since we set `saveat=0.1`. If we did not set `saveat`,
+this would be adaptively chosen. More details on handling the solution type
+[can be found in the DifferentialEquations.jl documentation](http://docs.juliadiffeq.org/latest/basics/solution.html).
+
+One last thing to note is that we can make our initial condition (`u0`) and time spans (`tspans`)
+to be functions of the parameters (the elements of `p`). For example, we can define the `ODEProblem`:
+
+```julia
+u0_f(p,t0) = [p[2],p[4]]
+tspan_f(p) = (0.0,10*p[4])
+p = [1.5,1.0,3.0,1.0]
+prob = ODEProblem(lotka_volterra,u0_f,tspan_f,p)
+```
+
+In this form, everything about the problem is determined by the parameter vector (`p`, referred to
+as `θ` in associated literature). The utility of this will be seen later.
+
+## Let's Put an ODE Into a Neural Net Framework!
+
+To understand embedding an ODE into a neural network, let's recall what a layer
+of a neural network actually is. A layer is a transformation 𝐑ⁿ→𝐑ᵐ which
+takes in a vector of size `n` and spits out a new vector of size `m`. That's it!
+So a differential equation layer is simply a layer that takes in a parameter
+vector `p` (of size `n`) and spits out a new vector (of size `m`). You can do this

```

Comment:
Well `p` can be the input to `u0` and final timespan, that is mentioned later down. So NeuralODE is only using the `p` here for its relation to `u0`, but in a sense it's an expanded form. It does interact oddly with the adjoint method though. @MikeInnes would you recommend having a separate input for these? I believe Flux will differentiate w.r.t. multiple different inputs? The best would be for it to be done via keyword arguments, i.e. `u0 = ..., p = ...`. Then NeuralODE would only use `u0` and other applications would use `p` or both.

Reply:
Sure but even if `p` is `u0` and final timespan it's weird to talk about the dimensionality of the *layer* being that of `p`, imo. Like in this example you wouldn't care about the dimensionality of the timespan.

Is there, in general, some other part of `p` that is changing through time besides `u`? Because when we talk about the layer dimensionality in any other layer we usually mean the dim of the input state, so `u0` and the dim of the output state, `u(t_end)`.  

I know you understand what I'm talking about, just being really explicit here. Is it worth it, in this blog, to be explicit about `p` dimensionality. Or is that a more niche interest? Could we keep it consistent with NeuralODEs in this blog for comparability if that's a reasonable simplifying assumption for an introductory blog?

Reply:
Is this just a naming thing? What about just calling it `x` instead and noting that `x` is essentially `x = [u0,p,tspan]`, would that make it more clear? Dealing with multiple inputs is probably a better way to handle it if Flux can do that.

>I know you understand what I'm talking about, just being really explicit here. Is it worth it, in this blog, to be explicit about p dimensionality. Or is that a more niche interest?

No, thinking of the solution of an ODE as a function of `p` is definitely not of niche interest. It's the basis of parameter estimation which is by far the biggest use of optimization on ODE solutions. Actually, NeuralODE is very niche in comparison to parameter estimation (which is where the adjoint methods originate from). So for the purposes of integrating data science and ML with differential equations, having the diffeq as both a function of `u0` and `p` are vital to the standard use cases.

Reply:
The confusion here is because Jesse is thinking of optimisation parameters, and Chris is thinking of parameters of the ODE. Probably in all of the neural ODEs work they were the same, but in e.g. the lotka-volterra case they are not; the NN actually generates a vector `p` that gets fed into `f(p)` (which happens to solve an ODE, but isn't a neural ODE in the sense of the paper). In that sense `p` is data in the network rather than an optimisation parameter.

I'm not sure how best to resolve this; one option is to clearly distinguish upfront between the two (and it might be best to describe optimisation parameters as "weights"). I think it's worth spending some time to get this right though, or it'll confuse a lot of people.

Reply:
Definitely good to be explicit upfront about the difference; it confused me too.